### DISCUSSION TITLE
Stacks-node on Raspberry PI

### DISCUSSION THREAD

**Discussion by laststrawman** [stacks-network/stacks-core#3136] [Q&A]
I am following these directions ... https://docs.stacks.co/en/nodes-and-miners/running-mainnet-node
I am getting the following error ... 
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
5ba6831a64933d47677db40792463843105e10ae6eb0338ecefe707b83cd887c

Anyone have any suggestions?

**Comment by wileyj** [comment]
the only options are to build the image locally on your pi, or use these images that i made *for* a pi (these won't be updated regularly, but you can use for testing purposes):
https://hub.docker.com/repository/docker/wileyj/stacks-blockchain-api
https://hub.docker.com/repository/docker/wileyj/stacks-blockchain

**Comment by zone117x** [comment]
You could also download the arm64 binary [`linux-arm64.zip`](https://github.com/stacks-network/stacks-blockchain/releases/download/2.05.0.2.0/linux-arm64.zip) from the release assets here https://github.com/stacks-network/stacks-blockchain/releases/tag/2.05.0.2.0 

@wileyj arm64 docker images should be built on release, see https://github.com/stacks-network/stacks-blockchain/blob/master/.github/workflows/docker-platforms.yml
Any idea what's going on there?

**Reply by wileyj** [reply]
> Any idea what's going on there?

None - i noticed the missing arm images myself last week (someone else was asking about them), so i just built them on a local pi for their use. 
probably worthwhile to get the workflow building these, particularly since there's at least 1 wishlist item i'm aware of targeting pi's.

**Reply by ceramicwhite** [reply]
It's been like this since mainnet launch. I kept praying we would get an Arm build on docker hub someday, thought there was a reason behind this not being done as part of releases

**Comment by laststrawman** [comment]
I am wanting to run a validating node like I do for the BTC network through Umbrel on the PI. I will keep you posted as to my progress.

**Reply by wileyj** [reply]
talk to @ceramicwhite - i was helping him with that specifically. 
there are some things on the umbrel setup you'll need to adjust in order for things to come up properly (while using the umbrel bitcoin service as the burnchain). 

https://github.com/stacks-network/stacks-blockchain-docker/tree/jasper

is the branch i made for him to get this working (he still had to adjust his config, but i'm pretty sure it was syncing on his umbrel node). 
be aware that you shouldn't do this unless you have the 8GB memory pi. otherwise you'll definitely hit OOM issues.

**Reply by ceramicwhite** [reply]
As wileyj recommended use that branch

These are the steps to get a Stacks Node/Api working on your Umbrel using the bitcoind from the Umbrel compose stack.

```
git clone --single-branch https://github.com/stacks-network/stacks-blockchain-docker -b jasper && cd stacks-blockchain-docker

cp sample.env .env

nano conf/mainnet/Config.toml.sample

# Update the Burnchain details
[burnchain]
chain = "bitcoin"
mode = "mainnet"
peer_host = "10.21.21.8"  <--- IP to Bitcoind on Umbrel (This is default, yours will be the same) 
username = "umbrel"       <--- umbrel rpc login (This default, yours will be the same)
password = "rpc_password"  <------ Go on your Umbrel Dashboard go to "Connect Wallet" at the bottom select "Bitcoin Core RPC" put the password here
rpc_port = 8332
peer_port = 8333

./manage.sh -n mainnet -a start
```
That's it!
One of my Wishlists grants is a one line script to "inject" the Stacks Node/API & the Explorer into the Umbrel App store someone is working on this currently, so will be easier in the future.

**Reply by laststrawman** [reply]
Looks like I am running ... Can I tail a log? Or is that what this command does? Follow logs: manage.sh -n mainnet -a logs

**Reply by ceramicwhite** [reply]
yeah that's what the log flag does but for all three containers(node/api/postgres) you could also just use docker logs to focus on one container or use this simple script @wileyj recommended to keep tabs on the nodes height as it syncs:

```
sleep=10
while true; do
    echo
    echo "3999      : $(curl -sL localhost:3999 | jq .chain_tip.block_height)"
    echo "stacks api: $(curl -sL api.stacks.org | jq .chain_tip.block_height)"
    echo "sleep $sleep"
    sleep $sleep
done
```

**Reply by laststrawman** [reply]
Hmmm ... not sure I am running ... ERROR: Service "bns" uses an undefined network "stacks-blockchain

**Reply by ceramicwhite** [reply]
> Hmmm ... not sure I am running ... ERROR: Service "bns" uses an undefined network "stacks-blockchain

Ahh I see what's going on. 
It's probably still running though just the log flag from the script is throwing an error as the docker network has been changed to default aka umbrel but we didn't change it in the compose files of bns.yaml and proxy.yaml but it wouldn't have mattered when you used the script to start the node(Just tried it)

Do `sudo docker ps -a` and check what the status is on stacks-blockchain container

Then do `sudo docker logs stacks-blockchain -f` for it's logs

If it's running run this: 
`curl -sL localhost:20443/v2/info | jq`   what does "stacks_tip_height" list?
or query just the tip with `curl -sL localhost:3999 | jq .chain_tip.block_height`

I'll send a PR for the BNS and proxy anyways

**Reply by laststrawman** [reply]
No joy ... looking at the docker logs ... Dumping core for pid 1
I think it might be using the wrong config ...
INFO [1652920360.452796] [testnet/stacks-node/src/main.rs:82] [main] stacks-node No Version Info (No Branch Info:No Commit Info+, release build, linux [aarch64])
INFO [1652920360.452857] [testnet/stacks-node/src/main.rs:115] [main] Loading config at path /src/stacks-blockchain/Config.toml

**Reply by wileyj** [reply]
> Hmmm ... not sure I am running ... ERROR: Service "bns" uses an undefined network "stacks-blockchain

fixed - those compose files weren't part of the updates for this branch. the networks were still the same as `master` branch - updated to `default` so they should work fine now.

**Reply by wileyj** [reply]
> No joy ... looking at the docker logs ... Dumping core for pid 1 I think it might be using the wrong config ... INFO [1652920360.452796] [testnet/stacks-node/src/main.rs:82] [main] stacks-node No Version Info (No Branch Info:No Commit Info+, release build, linux [aarch64]) INFO [1652920360.452857] [testnet/stacks-node/src/main.rs:115] [main] Loading config at path /src/stacks-blockchain/Config.toml


However you tried to run this, it's best to back out. stop all stacks-blockchain services (if running). 
```
umbrel@umbrel:~/stacks-blockchain-docker $ ./manage.sh -n mainnet -a stop
Stopping all services for stacks-blockchain 
Stopping all services 
Removing stacks-blockchain     ... done
Removing stacks-blockchain-api ... done
Removing postgres              ... done
Removing network umbrel_main_network
ERROR: error while removing network: network umbrel_main_network id 6ed737adaad33b3b8b982e60e184abd9921b7b48ba698e7a84bef05ab09d5a1f has active endpoints
```
*the error above is normal and ok* - you may have to manually stop the containers, running this on umbrel is very much experimental and the scripts are adjusted yet to recognize it's on an umbrel. 

next, run these commands **in order**:
```
$ cd
$ git clone --single-branch https://github.com/stacks-network/stacks-blockchain-docker -b jasper && cd stacks-blockchain-docker
$ export BTC_HOST="10.21.21.8"
$ export BTC_RPC_USER="umbrel"
$ export BTC_RPC_PASS="<umbrel btc rpc password>"
$ sed -i "
    /^peer_host/s/.*/peer_host = \"${BTC_HOST}\"/; \
    /^username/s/.*/username = \"${BTC_RPC_USER}\"/; \
    /^password/s/.*/password = \"${BTC_RPC_PASS}\"/; \
" ./conf/mainnet/Config.toml.sample
```
then, you can start the services normally. it looks like you tried to download BNS data first, in which case you'll want to try that again and wait for it to complete (as well as uncomment the `#BNS_IMPORT_DIR` in `.env`:
```
$ ./manage.sh -n mainnet -a bns
```

```
3umbrel@umbrel:~/stacks-blockchain-docker $ ./manage.sh -n mainnet -a status

Stacks Blockchain services are not running 

umbrel@umbrel:~/stacks-blockchain-docker $ ./manage.sh -n mainnet -a start
Creating persistent-data for mainnet 
Starting all services for stacks-blockchain 
Creating postgres ... done
Creating stacks-blockchain-api ... done
Creating stacks-blockchain     ... done
[ Success ]      Brought up mainnet
    Follow logs: ./manage.sh -n mainnet -a logs 
umbrel@umbrel:~/stacks-blockchain-docker $ ./manage.sh -n mainnet -a logs
Attaching to stacks-blockchain, stacks-blockchain-api, postgres
stacks-blockchain-api    | {"level":"info","message":"PgNotifier connected, listening on channel: stacks-api-pg-notifier","timestamp":"2022-05-19T02:19:30.990Z"}
stacks-blockchain-api    | {"level":"info","message":"Stacks 1.0 token offering data import started","timestamp":"2022-05-19T02:19:31.140Z"}
stacks-blockchain        | INFO [1652926761.221409] [testnet/stacks-node/src/main.rs:82] [main] stacks-node No Version Info (No Branch Info:No Commit Info+, release build, linux [aarch64])
stacks-blockchain        | INFO [1652926761.226329] [testnet/stacks-node/src/main.rs:115] [main] Loading config at path /src/stacks-blockchain/Config.toml
stacks-blockchain        | INFO [1652926761.355442] [testnet/stacks-node/src/event_dispatcher.rs:952] [main] Registering event observer at: stacks-blockchain-api:3700
stacks-blockchain        | WARN [1652926761.446735] [src/chainstate/coordinator/mod.rs:919] [main] Sortition DB /root/stacks-blockchain/data/mainnet/burnchain/sortition does not exist; assuming it will be instantiated with the correct version
stacks-blockchain        | WARN [1652926761.447368] [src/chainstate/coordinator/mod.rs:935] [main] Chainstate DB /root/stacks-blockchain/data/mainnet/chainstate does not exist; assuming it will be instantiated with the correct version
stacks-blockchain        | INFO [1652926761.447840] [testnet/stacks-node/src/run_loop/neon.rs:348] [main] Start syncing Bitcoin headers, feel free to grab a cup of coffee, this can take a while
stacks-blockchain        | INFO [1652926762.695219] [src/burnchains/bitcoin/spv.rs:910] [main] Syncing Bitcoin headers: 0.3% (2000 out of 736988)
stacks-blockchain        | INFO [1652926764.987015] [src/burnchains/bitcoin/spv.rs:910] [main] Syncing Bitcoin headers: 0.5% (4000 out of 736988)
stacks-blockchain        | INFO [1652926766.509789] [src/burnchains/bitcoin/spv.rs:910] [main] Syncing Bitcoin headers: 0.8% (6000 out of 736988)
...
```

**Reply by ceramicwhite** [reply]
Much cleaner. Love it!

**Reply by laststrawman** [reply]
Ha! Success! Now my little PI is validating the BTC network and the STX network! ... I think. Tailing the logs I see "Syncing Bitcoin headers: 56.2% blah blah blah". That completed and now I am getting a lot of testnet connection refused messages Event dispatcher

**Reply by laststrawman** [reply]
stacks-blockchain        | WARN [1652957125.503692] [testnet/stacks-node/src/event_dispatcher.rs:123] [chains-coordinator] Event dispatcher: connection failed  - Os { code: 111, kind: ConnectionRefused, message: "Connection refused" }
^CERROR: Aborting.

**Reply by wileyj** [reply]
> stacks-blockchain | WARN [1652957125.503692] [testnet/stacks-node/src/event_dispatcher.rs:123] [chains-coordinator] Event dispatcher: connection failed - Os { code: 111, kind: ConnectionRefused, message: "Connection refused" } ^CERROR: Aborting.

last reply 4 hours ago - now you start running commands against the API, i.e. `curl -sL localhost:3999` from your umbrel.
https://hirosystems.github.io/stacks-blockchain-api/

**Reply by laststrawman** [reply]
Block tip height 11342

**Reply by wileyj** [reply]
> Block tip height 11342

that's it - you have all the services running now

**Reply by laststrawman** [reply]
Soooo I have to ask ... How far am I from being able to mine from the PI? Or do I have the ability right now?

**Reply by laststrawman** [reply]
A little update ... 1. The BTC node running under Umbrel mounts their data on an external drive. If we ever get around to a stacks man on how to install we should do the same as I ran out of space on /dev/root and then had to create a symbolic link to persistent-data on the external drive (yeah I should have thought of it). 2. My poor little PI is running hot. Running top it seems to be stacks-no+ and postgres that seem to be hogging CPU.

**Comment by wileyj** [comment]
you do not. using umbrel for a miner would be very difficult, and would
involve forking umbrel and making updates to how it works. for stacks
mining, bitcoin needs to respond on 127.0.0.1 meaning your bitcoin
container would also need to run your stacks blockchain.

umbrel is a bad choice for a miner, your better off running binaries of btc
and stacks.

**Reply by laststrawman** [reply]
Ok ... well I accomplished what I set out to do and that is to just run a node that validates the network. Thanks for all your help.

**Comment by wileyj** [comment]
if you're using the repo i sent previously, yes - that's the command to
tail logs
